homeassistant/components/climate/__init__.py

```diff
@@ -404,7 +404,13 @@ def _report_turn_on_off(feature: str, method: str) -> None:
                 ClimateEntityFeature.TURN_ON
             )
 
-        if (modes := self.hvac_modes) and len(modes) >= 2 and HVACMode.OFF in modes:
+        if (
+            (modes := self.hvac_modes)
+            and len(modes) >= 2
+            and HVACMode.OFF in modes
+            and not supported_features & ClimateEntityFeature.TURN_ON
+            and not supported_features & ClimateEntityFeature.TURN_OFF
+        ):
```

Comment:
I think my previous review was incorrect.

However, I also think the check for the feature flags should be moved up:

```diff
         # Adds ClimateEntityFeature.TURN_OFF/TURN_ON depending on service calls implemented
         # This should be removed in 2025.1.
         if self._enable_turn_on_off_backwards_compatibility is False:
             # Return if integration has migrated already
             return

         supported_features = self.supported_features
+        if supported_features & (
+            ClimateEntityFeature.TURN_ON | ClimateEntityFeature.TURN_ON
+        ):
+            # The entity supports both turn_on and turn_off, the backwards compatibility
+            # checks are not needed
+            return
+
         if not supported_features & ClimateEntityFeature.TURN_OFF and (
```

Also, we should add a test which passes with the changes in this PR and fails without it

Reply:
We implemented the `_enable_turn_on_off_backwards_compatibility` flag because supported features are dynamic and could be added in a later stage than during initialization.
But I assume the above suggestion by @emontnemery  would be fine as if the integration from the init has the flags we know they are supported and don't need to continue to check.